rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,170 @@
+
+### `lookup_csv_rows` 
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the 
+array.
+
+### Parameters 
+
+*filename*
+
+The filename is the path to the filename of the csv file to lookup. This
+filename has to be a static string (cannot be changed at runtime) so that 
+Vector can:
+
+1. Validate that the file exists.
+2. Load the file into memory.
+
+The filename will be relative to the current working directory that Vector is
+running in.
+
+The CSV file must start with a title row that gives a name to each of the 
+columns provided in the file.
+
+*key*
+
+The CSV file may be encrypted. This parameter specifies a key to use to decrypt
+the file.
+
+Is it safe to assume/require the file to be encrypted using AES, or should we
+support other algorithms as well?
+
+*criteria*
+
+`criteria` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_insensitive*
+
+By default the search will be case sensitive, but this can be changed by 
+passing `true` to this parameter.
+
+## Internal Proposal
+
+- Describe your change as if you were presenting it to the Vector team.
+- Use lists, examples, and code blocks for efficient reading.
+- Be specific!
+
+The majority of the changes will be provided by adding two new functions within
+`vrl/stdlib` and some shared lookup code.
+
+Since the CSV file is loaded at bootime, it is not essential that it is loaded 
+using async. This makes things much simpler as currently the VRL compilation 
+process is entirely synchronous. No code outside of the new function in the
+vrl stdlib should need changing.
+
+The entire data file will be loaded into memory, so all lookups will be 
+performed in memory.
+
+### Indexing 
+
+Although the first version is not going to be doing any indexing into the data
+it is worth bearing it in mind as we will most likely need to add this in due
+course.
+
+In order to perform the indexing VRL needs to know which fields to index. The 
+criteria is being passed in as an object. If the type def for that object is
+known at runtime, we can extract the fields from this. 
+
+If the type def isn't known we could 
+
+- raise an error 
+- perform the search unindexed
+- create the index at runtime - generally the shape of this object will not 
+  change over time, so most likely the cost of creating the index would only
+  need to be paid on the first event.
+
+The approach we take here should be decided before the first release so we don't
+have to introduce a breaking change further down the line.
+
+Actual indexing strategies can be decided later.
+
+## Rationale
+
+There is significant customer demand for this feature.
+
+## Drawbacks
+
+1. There are a number of issues with parsing CSV files in particular around
+   handling delimeters and separators.
```

Comment:
I was thinking the same. I'm pretty sure we'll need a function argument to allow defining the delimiters because `,` and `;` are both very common.

Reply:
Yeah, we should probably support the same parameters that `parse_csv` does.